### ISSUE TITLE
[audioworklet] AudioWorklets integration with WebAssembly

### ISSUE THREAD

**Issue by juj** [WebAudio/web-audio-api-v2#5]
Reading the AudioWorklet reference at https://github.com/WebAudio/web-audio-api/wiki/AudioWorklet-Examples and https://github.com/WebAudio/web-audio-api/wiki/AudioWorklet-IDL-Proposal, it looks like code that would utilize WebAssembly to render audio in an AudioWorklet would look something like the following:

```javascript
registerAudioWorkletProcessor('WasmInAudioWorklet', class extends AudioWorkletProcessor {
  constructor (options) {
    super(options);

    // Compile and Instantiate Wasm Module here, communicate with SharedArrayBuffer with the remaining program
    this.wasmModule = ...;
    this.wasmInstance = ...;
    this.wasmHeap = new Float32Array(...);
  }

  process(inputs, outputs, parameters) {
    for(var i = 0; i < inputs.length; ++i) { // *
      var input = inputs[i]; // *
      for(var j = 0; j < input.length; ++j) this.wasmHeap[... + j] = input[j]; // *
    } // *
    this.wasmInstance.processAudio();

    for(var i = 0; i < outputs.length; ++i) {  // *
      var output = outputs[i]; // *
      for(var j = 0; j < output.length; ++j) { // *
        output[j] = this.wasmHeap[... + j]; // *
      } // *
    } // *
  }
});
```

Talking with WebAssembly developers and @lukewagner, he is suggesting that it would be made possible to register the WebAssembly heap at AudioWorklet construction time, so that one would be able to optimize away the copies on the lines marked with stars `// *`.

Would something like this be feasible?

**Comment by karlt** [comment]
An extension point to consider when designing new APIs is an alternative `registerProcessorV2()` method on AudioWorkletGlobalScope.  From a wasm perspective, registering a function would be more natural than a class.

**Comment by guest271314** [comment]
Since `Response` https://developer.mozilla.org/en-US/docs/Web/API/Response is not defined in `AudioWorkletGlobalScope` are `WebAssembly.compileStreaming()` https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WebAssembly/compileStreaming and `WebAssembly.instantiateStreaming()` https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WebAssembly/instantiateStreaming expected to not work and are users of `AudioWorklet` expected to know that those methods will not work due to no `Response` object being define even though the methods are still defined on `WebAssembly` object in `AudioWorkletGlobalScope`?

```
WebAssembly {compile: ƒ, validate: ƒ, instantiate: ƒ, compileStreaming: ƒ, instantiateStreaming: ƒ, …} bypass-processor.js:18 {numberOfInputs: 1, numberOfOutputs: 1}
```

**Comment by rtoy** [comment]
Teleconf: Profiling the cost of copying to and from the WASM heap would be quite interesting.

**Comment by guest271314** [comment]
@rtoy Currently working on exactly that use case. Due to Chromium refusing to support capture of monitor devices at `getUserMedia({audio: true})` and Web Audio API refusing to support `fetch()` in `AudioWorkletGlobalScope` created several workarounds which use Native Messaging and Native File System to stream system audio output capture to `AudioWorklet` which, if necessary is then connected to a `MediaStreamTrack` (`MediaStreamAudioDestinationNode`) and recorded with `MediaRecorder`, see https://bugs.chromium.org/p/chromium/issues/detail?id=1115640 for the current state and how, with the cooperation of the relevant parties, numerous use cases can be achieved.

One issue is that there are two different browsing contexts. Another issue is `WebAssembly.Memory.grow()` https://github.com/WebAssembly/design/issues/1358 was evidently not developed in cooperation from JavaScript (Ecmascript; TC39) https://github.com/tc39/test262/issues/2719, thus, it is difficult to actually start with `new WebAssembly.Memory({initial:1, maximum:65536 * 3 * 60 * 60) / 65536, // 1 hour of 2 channel PCM })`  and call `grow(N)` when the `ReadableStream` from one or more `fetch()` calls or AsyncGenerator is reading one or more files written to local filesystem by native application or shell script.

It would be ideal to be able to write directly to a single WebAssembly Memory instance within the shell script which is the same Memory object exported to JavaScript, however, WASI is not obliged to have JavaScript be on-topic at all https://github.com/WebAssembly/WASI/issues/307.

It should be as simple as using PHP `passthru()` to capture any device ("headphones", "speakers", "speech-dispatcher", or any virtual device the user might create or set to on at the kernel level). This actually can do what we want

```
<?php 
  if (isset($_GET["start"])) {
    header("Access-Control-Allow-Origin: *");
    header("Content-Type: application/octet-stream");
    echo passthru("parec -v --raw -d alsa_output.pci-0000_00_1b.0.analog-stereo.monitor");
    exit();
  }
```
```

   let controller, signal;
   const [start, stop] = document.querySelectorAll("button");
   start
   .onclick = async e => {
        start.disabled = true;
        // set parameters as arrays for potential "infinite" input, output stream
        let {track, port} = await new AudioWorkletStream({ 
          latencyHint: 1,
          workletOptions: {
            numberOfInputs: 1,
            numberOfOutputs: 2,
            channelCount: 2,
            processorOptions: {
              codec: 'audio/wav',
              offset: 0
            },
          },
        }).promise;

        console.log(track, port);
     controller = new AbortController();
     signal = controller.signal;
     fetch("http://localhost:8000?start=true", {signal})
     .then(r => r.body)
     .then(readable => {
       const reader = readable.getReader();
       reader.read().then(function process({value, done}) {
         if (done) {
           console.log(done);
           return;
         }
         console.log(value);
         return reader.read().then(process);
       })
       .catch(console.error);
     })
   }
   stop.onclick = e => {
     controller.abort();
   }
```

however, it is not a simple matter to write that data to a single memory object in PHP realm than is exposed to JavaScript, where instead of reading from the `ReadableStream` then writing to `SharedArrayBuffer`, we should just be able to write to `SharedArrayBuffer` in PHP, Jelly, R, Rust, bash, etc., then just read that data in JavaScript. `Map`s can store 16 million elements, yet might have issues with regard to reading that data re glitches and gaps in playback.`TypedArray`s do not currently grow, need to handle `RangeError`s. Native File System does not currently support reading from a named pipe or a file being written to in parallel without reading the entire file at each iteration of a polling loop, while handling errors. Native Messaging is non-active specification, and it is not clear if deprecation of Chromium apps means deprecation of Native Messaging, with which have been able to achieve use cases not possible in the given browser source code alone.

Am willing to test experiments and proof-of-concepts, as do anyway, thoroughly.

**Comment by guest271314** [comment]
The PHP and JavaScript code in the above is another WIP, and adds the networking element, CORS, etc. The linked proposal would essentially be the same output as the `fetch()` and `passthru()`, without unnecessary restrictions. Not trying to squeeze existing API's into use cases where not possible. We should be able to communicate directly with a shell script or native application and for STDIN and STDOUT to be written to a single memory allocation instance (WebAssembly.Memory; WASM) that dynamically grows when STDOUT is dynamic (e.g., a media stream) that is streamed using a `TransformStream` to and from, something like https://stackoverflow.com/a/2789967, where when the stream is closed the memory is flushed. 

It is difficult to fathom the maintainers of Transferable Streams, WASI, WebAssembly, Native File System, Web Audio API cannot achieve any clear and concise goal that is set.

**Comment by guest271314** [comment]
FWIW, streaming captured system audio output from `fetch()` in Chromium background script after starting a local server to passthrough raw PCM to an arbitrary web page, transfering the `ReadableStream` where text is converted to `TypedArray` to `AudioWorkletProcessor` where `Uint8Array` are written to `WebAssembly.Memory` instance with `shared` set to `true` and `grow(1)` is executed to dynamically grow memory during the stream. One limitation at Chromium appears to be maximum size of underlying memory set by `grow()` https://github.com/wasmerio/php-ext-wasm/issues/121#issuecomment-678898047.

**Comment by padenot** [comment]
> An extension point to consider when designing new APIs is an alternative `registerProcessorV2()` method on AudioWorkletGlobalScope. From a wasm perspective, registering a function would be more natural than a class.

It can be made to work without changing the name of the function, and instead dispatching based on the shape of the second argument. If it's an ES6 class, it has to conform to the required prototype.

If it's an object, it needs to have the right keys, and those keys need to be functions. They can have the same names, and be implemented in js or directly in WASM. This clealry ties in to #4. When #4 is solved, then I believe the solution for #5 will come naturally.

**Comment by carlsmith** [comment]
The original example above contains this block:

``` js
// Compile and Instantiate Wasm Module here, communicate with SharedArrayBuffer with the remaining program
    this.wasmModule = ...;
    this.wasmInstance = ...;
    this.wasmHeap = new Float32Array(...);
```

In practice, you cannot instantiate a Wasm module in a constructor like that, as instantiation is asynchronous, so the exports must be assigned to instance variables in a callback (after the constructor has returned, and possibly after the `process` method attempts to call an exported function). The issue linked to immediately above addresses that problem specifically.

**Comment by mdjp** [comment]
This would be a great issue to discuss across groups as TPAC 2021

**Comment by guest271314** [comment]
Implementation of `WebAssembly.Mempory.grow()` with `AudioWorklet` https://github.com/guest271314/webtransport/blob/main/webTransportAudioWorkletWebAssemblyMemoryGrow.js. Note, when using a `ReadableStream` to supply input data concurrently with writing data to `AudioWorklet` (or `MediaStreamTrackGenerator` https://github.com/guest271314/native-messaging-espeak-ng/blob/master/AudioStream.js) the first few seconds of audio output can contain glitches due to fragmentation, and/or the same thread not being able to process more than on stream at one time. To avoid that issue the fetching of input data and processing of raw audio can (should) be farmed out to one or more `Worker` threads or `ServiceWorker` instances.
